kratos/utilities/mortar_utilities.h

```diff
@@ -602,12 +610,282 @@ class MortarUtilities
+        Variable<array_1d<double, 3>>& ThisVariable
+        )
+{
+    auto& aux_value = ThisNode.GetValue(ThisVariable);
+    aux_value = ZeroVector(3);
+}
+
+template<> 
+inline void MortarUtilities::MatrixValue<Variable<double>, Historical>(
+        GeometryType& ThisGeometry,
+        Variable<double>& ThisVariable,
+        Matrix& ThisValue
+        )
+{
+    ThisValue.resize(ThisGeometry.size(), 1, false);
+    for (unsigned int i_node = 0; i_node < ThisGeometry.size(); i_node++)
+    {
+        ThisValue(i_node, 0) = ThisGeometry[i_node].FastGetSolutionStepValue(ThisVariable);
+    }
+}
+
+template<> 
+inline void MortarUtilities::MatrixValue<component_type, Historical>(
+        GeometryType& ThisGeometry,
+        component_type& ThisVariable,
+        Matrix& ThisValue
+        )
+{
+    ThisValue.resize(ThisGeometry.size(), 1, false);
+    for (unsigned int i_node = 0; i_node < ThisGeometry.size(); i_node++)
+    {
+        ThisValue(i_node, 0) = ThisGeometry[i_node].FastGetSolutionStepValue(ThisVariable);
+    }
+}
+
+template<> 
+inline void MortarUtilities::MatrixValue<Variable<array_1d<double, 3>>, Historical>(
+        GeometryType& ThisGeometry,
+        Variable<array_1d<double, 3>>& ThisVariable,
+        Matrix& ThisValue
+        )
+{
+    ThisValue.resize(ThisGeometry.size(), ThisGeometry.WorkingSpaceDimension(), false);
+    for (unsigned int i_node = 0; i_node < ThisGeometry.size(); i_node++)
+    {
+        row(ThisValue, i_node) = ThisGeometry[i_node].FastGetSolutionStepValue(ThisVariable);
+    }
+}
+template<> 
+inline void MortarUtilities::MatrixValue<Variable<double>, NonHistorical>(
+        GeometryType& ThisGeometry,
+        Variable<double>& ThisVariable,
+        Matrix& ThisValue
+        )
+{
+    ThisValue.resize(ThisGeometry.size(), 1, false);
+    for (unsigned int i_node = 0; i_node < ThisGeometry.size(); i_node++)
+    {
+        ThisValue(i_node, 0) = ThisGeometry[i_node].GetValue(ThisVariable);
+    }
+}
+
+template<> 
+inline void MortarUtilities::MatrixValue<component_type, NonHistorical>(
+        GeometryType& ThisGeometry,
+        component_type& ThisVariable,
+        Matrix& ThisValue
+        )
+{
+    ThisValue.resize(ThisGeometry.size(), 1, false);
+    for (unsigned int i_node = 0; i_node < ThisGeometry.size(); i_node++)
+    {
+        ThisValue(i_node, 0) = ThisGeometry[i_node].GetValue(ThisVariable);
+    }
+}
+
+template<> 
+inline void MortarUtilities::MatrixValue<Variable<array_1d<double, 3>>, NonHistorical>(
+        GeometryType& ThisGeometry,
+        Variable<array_1d<double, 3>>& ThisVariable,
+        Matrix& ThisValue
+        )
+{
+    ThisValue.resize(ThisGeometry.size(), ThisGeometry.WorkingSpaceDimension(), false);
+    for (unsigned int i_node = 0; i_node < ThisGeometry.size(); i_node++)
+    {
+        row(ThisValue, i_node) = ThisGeometry[i_node].GetValue(ThisVariable);
+    }
+}
+
+template<> 
+inline void MortarUtilities::AddValue<Variable<double>, Historical>(
```

Comment:
How?

Reply:
#pragma omp barrier ?

Reply:
What i am telling is that either you call it AtomixAddValue and protect it by an atomic or you tell, at least in the doc, and better in the name, that it is not threadbare and it can not be done concurrently

Reply:
The process is standalone, so it will not use OMP if not modified explicitily

Reply:
What i am suggesting is exactly this. A user does not know at first sight if this can be used in a openmp loop or not. My suggestion is just to put a comment telling if it is or not theadsafe (the answer appears to be NOT)